Update validation for NLP tasks

osanseviero opened the pull request:
Summary of changes:
* Aligned max length with `max_new_tokens` in [docs](https://api-inference.huggingface.co/docs/python/html/detailed_parameters.html). 
* Changed `max_length` to `max_new_tokens` for `text-generation`, which I think would be the right param.
* Added `return_full_text` and `num_return_sequences` for `text-generation`
* Added `summarization`, `table-question-answering`, and `text2text-generation` validation.
* Fixed `conversational`. It was using params instead of inputs, and the inputs are others.

Let me know if you disagree with any and I can revert :)

Narsil commented on api-inference-community/tests/test_nlp.py:
```diff
@@ -49,16 +49,149 @@ def test_sentence_similarity(self):
+        self.assertEqual(processed_params, params)
+        self.assertEqual(normalized_inputs, "whatever")
+
+    @parameterized.expand(
```
I find those quite hard to read and quite error-prone. I've seen quite a few mistakes in transformers tests by over parametrization.

What's this issue with simply writing those explicitly ?
Another problem with the current implementation is that it's impossible to test parameters coupling ('min_length '> 'max_length' for instance.

osanseviero replied:
Fixed. I also added a test for the last point you mentioned.

Narsil commented on api-inference-community/api_inference_community/validation.py:
```diff
@@ -36,48 +48,69 @@ class Temperature(ConstrainedFloat):
+
+
+class TextGenerationParamsCheck(BaseModel):
+    max_new_tokens: Optional[MaxLength] = None
```
This should be MaxNewTokens which is different from MaxLength imo, max_length has it's purpose if you want to avoid go beyond the model capacity, so it still has some purpose.

Also I'm never a super fan of breaking code symmetry.

osanseviero replied:
Fixed.

Narsil reviewed (changes requested):
I'm fine with enabling max_new_tokens by default instead of max_length however, I think that it should get its own validation class.

For conversation, we were missing the input validation for sure, but the parameters still do exist (but they are NOT the same as text anymore though, return_num_sequences is impossible for instance, at least in transformers.)

I know it's cumbersome to write many repetitive tests, but overall I think it works better than parametrizing, as parametrization leaves more room for silently failing tests, and reading tests when they actually are failing is sometimes hard to untangle and update (especially when the tests you want to modify need to significantly differ from what they are tied to).

ofc all this is a taste matter, so consider the remarks about testing as NITs

osanseviero commented:
For `conversation`, I don't see any params in [https://api-inference.huggingface.co/docs/python/html/detailed_parameters.html#conversational-task]. I'm trying to keep them in sync. As far as I understand, all of these seq2seq models call a `generate` function in Transformers, so I'm not sure which would be the eligible params for `conversation`.

Narsil commented:
`top_p`,`top_k`, `repetition_penalty`, `max_time` etc... are valid. 

You are correct, it's using `generate`, it's either full decoder (DialoGPT-large) or encoder-decoder (Blenderbot).

The only reason it's not entirely the `text-generation` parameters, it's because of the return type which is a dictionary and so `num_return_sequences` would require returning a list.
We definitely can add them to the docs.

osanseviero commented:
Updates
* Changed to a shared params class for the generation ones with specific params in subclasses
* Reverted max_tokens to be consistent with the 3 types.
* Added param validation for conversational.
* De-parameterized some of the tests
* Added validation for `max_length` > `min_length`.

I'll send another PR for updating docs

Narsil reviewed (approved):
LGTM
